File: py/portfolio.py
import functools
import operator
from money import Money


class Portfolio:
    def __init__(self):
        self.moneys = []

    def add(self, *moneys):
        # should have another look at common methods for iterables..."extend list by appending elements from iterable"
        self.moneys.extend(moneys)

    def evaluate(self, currency):
        # map the moneys array to an array of all amounts, reduce the new array by adding; rather functional python approach
        total = functools.reduce(operator.add, map(lambda m: m.amount, self.moneys), 0)
        return Money(total, currency)

File: js/portfolio.js
const Money = require('./money');

class Portfolio {
	constructor() {
		this.moneys = [];
	}
	// rest parameter syntax allows a function to accept an indefinite number of arguments as an array
	add(...moneys) {
		this.moneys = this.moneys.concat(moneys);
	}

	evaluate(bank, currency) {
		let failures = [];
		let total = this.moneys.reduce((sum, money) => {
			try {
				let convertedMoney = bank.convert(money, currency);
				return sum + convertedMoney.amount;
			} catch (error) {
				failures.push(error.message);
				return sum;
			}
		}, 0);
		// no elements in an array <-> !array.length
		if (!failures.length) {
			return new Money(total, currency);
		}
		throw new Error('Missing exchange rate(s):[' + failures.join() + ']');
	}
}

module.exports = Portfolio;
